### ISSUE TITLE
[toggle][togglegroup] render function receives only part of the necessary props

### ISSUE THREAD

**Issue by dartess** [mui/base-ui#2255]
# Bug report

## Current behavior

The keyboard navigation for toggle group is broken if I use only those props that render function provides.

## Expected behavior

Props from render function should be enough for the work of keyboard navigation.

## Reproducible example

https://codesandbox.io/p/devbox/damp-wind-j72vfx?workspaceId=ws_8iqbh1gg1GcLonjY38LEti
(just focus on any item and try navigate through keyboard arrows)

## Base UI version

1.0.0-beta.1

## Which browser are you using?

Chrome

## Which OS are you using?

Mac OS

## Which assistive tech are you using (if applicable)?

Voiceover/JAWS etc.

## Additional context

I spent about two hours to figure out where the problem is. The element that is returned from the function is additionally added with props, such as another ref, some on-handlers and a tabindex. But I can't make a universal, independent and normally typed component because of this.

**Comment by atomiks** [comment]
The problem is Toggle (and other similar components) internally use `<CompositeItem render={element} />` and this requires _all_ incoming props from the component to be spread onto `ToggleRenderComponent`, like this:

```jsx
function ToggleRenderComponent(props: React.ComponentProps<"button">) {
  return (
    <button type="button" className={styles.Button} {...props}>
      <AlignLeftIcon className={styles.Icon} />
    </button>
  );
}

<Toggle value="left" render={<ToggleRenderComponent />} />
```

So it has two passes of rendering, and one of them is opaque/internal:

https://github.com/mui/base-ui/blob/3074ca21be68795bb6541be6fd974cfd0e35a02f/packages/react/src/toggle/Toggle.tsx#L70-L87

https://github.com/mui/base-ui/blob/3074ca21be68795bb6541be6fd974cfd0e35a02f/packages/react/src/composite/item/CompositeItem.tsx#L15-L18


---

But is there a reason this can't be used?

**Comment by dartess** [comment]
> this requires all incoming props from the component to be spread

Maybe I didn't have enough documentation and examples to understand this faster. 

I seem to have trouble understanding the logic behind how all this works for me as a user. In the documentation, I see that components accept render functions, and I see that props come as an argument. But then it turns out that these props are not enough. Why do they come at all then (if _all_ props can be implicitly)? How do I use all this? I would like to see more documentation and examples on this, so I don't get stuck when I e.g. get one `ref` from the props argument, and one implicit `ref`.

I want to have more clarity and less magic. In this case, I wanted to make a component that would not be tied to a specific `baseui` component, so I separately declared the `selected` prop (where I pass the `state.pressed`), and separately declared the `rootHtmlProps?: HTMLProps<HTMLButtonElement>` prop. In this form, I would not be tied to magic.

Afterall I was forced to rework the props to
```
type Props = HTMLProps<HTMLButtonElement> & {
  item: SelectItem;
  selected: boolean;
};
```
after which I have to use spread to separate my props from the props I have to throw at the root element, which looks fragile.

(I apologize for my wording, I'm just a little frustrated and having a hard time expressing my thoughts. I'm mostly in love with the baseui.)

**Comment by atomiks** [comment]
Yeah, I agree this is confusing 😓 - I was trying to find out why it wasn't working and how to get it working, but this should be fixed. The two render passes is not efficient to begin with
